### PR TITLE
RPC for scanning utxo set in pruned mode  

### PR DESCRIPTION
**Pull request by victorsk2019**
Good day!

This (draft) PR offers new RPC `getutxoforkey` for scanning utxo set and returning a utxo amount in pruned mode, as discussed.  It takes a private key and a block height required arguments.

This implementation synthesizes some features of already implemented code for `importprivkey` and `gettxoutsetinfo` and offers functionality that may be useful in situations for getting utxo amount in blocks that are outside of pruned data (and possibly extend this functionality to be able to import TX with utxo into a wallet from any block, because, I think, the ingredients for generating a TX for adding to wallet, like the presence of `CTxOut`, is already there).

I tried to implement efficient scans, ideally with O(1) running time, but the best I can come up with is O(nm), I believe.  Some member functions take TxId, so this information would need to be provided, but I think providing required block height information makes running time as optimal as possible: the longest I had to wait to find a utxo value on a Dell Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz was about 1 minute.  The CPU consumption with this implementation peaked to about 93%, so its probably within acceptable value?  

So, what it does is start by iterating over entire set and when specified block height is encountered, the check for address and its corresponding utxo is done.   Hence, by my analysis, the running time is O(nm) because both the entire set and a set of block heights is being iterated over in the worst case.

One thing I am still not clear about is the role of mutex thread locking: `LOCK(cs_main)` (line 1531 of wallet.cpp in this draft) and a call to `boost::this_thread::interruption_point()` (line 1543).  I understand the need for a mutex, I just don't quite understand it and the call to `interruption_point()` in that part yet 😙, so its something for me to figure out.

Cheers!
Victor.

### CODE REVIEW

**Comment by patricklodder** [comment]
First off, thanks for all the code. It's great to make some steps forward on this.

## Concept review

I'm going to describe the full concept as I understand it from past discussions, please correct me if I miss or misunderstand something. I will then compare this to the code proposed but only in terms of functionality, it's imho too early to do in-depth code review because I think some things should be changed. 

### Top level concept

I think that the ultimate (epic) use case this works towards can be formulated as:

> I, as a **pruned node user** want to **rescan my keys without having to download all blocks** to **spend the coin I own**

which is in principle the feature request from #3518. This is a really useful request, because it will _**significantly**_ reduce the load on the network.

As discussed previously in #3653, there is lower level functionality needed to achieve this, so this can be broken down into multiple smaller pieces of functionality.

### Breakdown

1. Create functionality for scanning the utxo database
2. (optionally) expose this functionality via RPC
3. Create a "rescan" process that can run a set of (HD and non-HD) `scriptPubKeys` provided by the wallet through the utxo database and inserts discovered records of spendable coin into the wallet if these didn't exist yet.
4. Expose rescan functionality through RPC
5. Expose rescan functionality in the Qt GUI

### Scope of this PR

This PR seems to implement point 1 and 2 from the breakdown, which is a very nice scope to start with - not too big to become hard to review, not too small to be meaningless. I'm not going to reverse-engineer describe the implementation but rather the concept, so there can (will) be differences between what's implemented and what I describe.

> I, as a **Dogecoin Core user**, want to **scan the utxo database for coin belonging to an output script**, to **get a fully self-validated view of which coin is unspent**

----

If I compare that to the approach taken in the code, I observe that the implementation proposes:

1. code in the wallet module but could (and imho should) be done outside this, because there is actually no integration with the wallet implemented or needed within this scope. I.e. the whole functionality is currently depending on `ENABLE_WALLET` and I think that utxo database scanning on its own is much more generic than that.

2. to scan a WIF formatted private key for utxo, which can be useful, but makes a lot of assumptions about the use of the key and perhaps a simpler yet more inclusive implementation (that would be easier to review and test) would be to scan addresses and/or `scriptPubKey`. For completeness, maybe 3 RPC endpoints should be provided: 
    - Scan p2pkh entries for a private key
    - Scan addresses (=p2pkh and p2sh)
    - Scan a script (=everything that at one point may have been allowed within policy but <is no longer/will be in the future>

this will then force the scanning routine to be generic, and take care of transformations (from address, from key) in callers of that routine?


3. a notion of scanning at a specific height. I think that scanning should not require height - it can be given as a hint though (don't bother before block x) but I don't think that this actually helps, especially because in the implementation the height is used as a filter that is applied within the main scan iteration, it does not reduce the number of iterations?

**Comment by victorsk2019** [comment]
Hi!

Thank you very much for the feedback.

>thanks for all the code. It's great to make some steps forward on this.

It's my pleasure.  Working with Dogecoin Core gives me the opportunity to directly apply what I learned in university C++ courses - it brings back a lot memories from many years ago!  Even though I've got 13 years as a Software Developer, that wasn't much related to university course work.

>a notion of scanning at a specific height. I think that scanning should not require height - it can be given as a hint though (don't bother before block x) 

I agree that as a standalone, this RPC with required height argument doesn't seem too useful because why would I want to know the utxo amount when I already know what the amount is at a given block height?  Though, there may be a situation if someone would want to check if running in a pruned mode given a private key and block height they can view their utxo amount.  However, my primary goal was to improve efficiency with scanning because applying a check for presence of address at each iteration seemed to be quite time consuming, so I've decided to optimize as much as possible by applying a check only when encountering specific block number (which has drastically reduced wait time to get the result).

But, the main reason for this RPC implementation, as I understand it, is to function as a "stepping stone" leading to implementation of importing private keys in pruned mode as a long-term goal.  My understanding is the major "stumbling block" in #3653 was inability to import keys for utxos in blocks that lie beyond pruned blocks.  So, for example, something like `importprivkeypruned` could later be implemented that uses functionalities in `getutxoforkey` as a basis.  For instance, `GetUTXOForPubKey()` in `getutxoforkey` would be equivalent to what `ScanForWalletTransactions()` does for non-pruned blocks case and then implement a pruned mode equivalent of `addToWallet()` in future `importprivkeypruned` RPC, using data obtain in `getutxoforkey`.  That's the main reason why I think utxo scanning function was included in `wallet.cpp` - to put similar functionalities in the same location.  Hence, the question I have is, can `getutxoforkey` functionality later be used for implementation of importing keys in pruned mode, or would it be the correct approach in this case?  Thanks!

**Comment by victorsk2019** [comment]
Note: this last commit scans backwards, starting from the end of utxo set.

**Comment by victorsk2019** [comment]
This last commit https://github.com/dogecoin/dogecoin/pull/3665/commits/c84e662de2e15784c167ff3074e07e1d0017f742 removes expensive process of extracting base58 address from `scriptPubKey`.  While reading over source code, I observe that class `CScript` seems to have an overloaded `==` operator making it possible to directly compare two `CScript` type variables without having to apply extra function calls to obtain string value, making scanning and retrieval of utxo value much faster.  In fact, during my tests, I now obtain utxo amount within 1 minute for test keys on my system 👍

File: qa/rpc-tests/blockchain.py [hunk 1]
```diff
@@ -112,6 +114,44 @@ def _test_gettxoutsetinfo(self):
...
+
+        # verify that txid, utxo amount and new block height for tx 
+        # with 42.001 Doges is the same as in this RPC response
+        assert_equal(res['amount'], Decimal("42.001"))
+        assert_equal(res['height'], 121)
+        assert_equal(res['txid'], txid)
+
+        # NOTE: calling self.sync_all() results in assertion failure
+        # 'Block sync to height 121 timed out' so as a workaround we
+        # currently check the presense of utxo only on node1
```
Comment (patricklodder) [inline qa/rpc-tests/blockchain.py:153]: Interesting. I'd normally look into the root cause but this is probably the wrong test class to integrate this into (as the other tests in this test assume a fixed state - it uses the premade chain and doesn't generate blocks anywhere).

Perhaps a standalone test class (and file) would be better.

**Comment by victorsk2019** [comment]
Just a minor correction for last commit: in RPC response from node0 we verify whether txid is the same as txid on node1.  It should read:
```
# After nodes have been syncronized, verify that txid, utxo amount and
# new block height for original tx on node1 with 42.001 Doges is the same 
# as in RPC response on node0 
```

**Comment by patricklodder** [comment]
Please let me know if you think this is ready for code review now. I will have some notes to improve this code.

**Comment by victorsk2019** [comment]
> Please let me know if you think this is ready for code review now. I will have some notes to improve this code.

Yes. I think it is ready for review now.  Other than adding more test cases not sure what else to improve here on my end 😉. 

The "bidirectional" scan I've implemented - to perform scan from either way at random - may seem a bit weird but in my opinion it makes the scan more fair so that utxos located in the second half of the set can have a chance to be scanned earlier.  Also, others may make use of the ability to scan both ways to improve on performance or may find it useful for other implementations.  Thanks.

**Comment by victorsk2019** [comment]
Some quick notes about https://github.com/dogecoin/dogecoin/pull/3665/commits/fa32550b61c7f5f7c35bdcc203d3b4ec3dcd33c9.  I've made RPC test in `utxo-scan.py` more flexible with optional number of random test amounts (I tested with maximum 100 test amounts) and tried to diversify test samples.  For some calls RPC response was `Unable to find utxo amount`.  I studied this case and from what I understand in some cases utxo amount was being used for input in one of the test transactions so that utxo amount no longer becomes available?
